drivers/include/mfrc522.h

```diff
@@ -0,0 +1,844 @@
+ * of pages 0Eh, 0Fh, 00h and 01h are returned.
+ *
+ * The buffer must be at least 18 bytes because a CRC_A is also returned.
+ * Checks the CRC_A before returning MFRC522_STATUS_OK.
+ *
+ * @param[in]      dev          Device descriptor of the MFRC522
+ * @param[in]      block_addr   MIFARE Classic: The block (0-0xff) number.
+ *                              MIFARE Ultralight: The first page to return data from.
+ * @param[out]     buffer       Buffer to store the data in
+ * @param[inout]   buffer_size  Buffer size, at least 18 bytes. Also number of
+ *                              bytes returned if MFRC522_STATUS_OK.
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
+ */
+mfrc522_status_code_t mfrc522_mifare_read(mfrc522_t *dev,
+                                          uint8_t block_addr,
+                                          uint8_t *buffer,
+                                          uint8_t *buffer_size);
+
+/**
+ * @brief Write 16 bytes to the active PICC
+ *
+ * @note  For MIFARE Classic the sector containing the block must be
+ *        authenticated before calling this function.
+ *
+ * For MIFARE Ultralight the operation is called "COMPATIBILITY WRITE". Even
+ * though 16 bytes are transferred to the Ultralight PICC, only the least
+ * significant 4 bytes (bytes 0 to 3) are written to the specified address. It
+ * is recommended to set the remaining bytes 04h to 0Fh to all logic 0.
+ *
+ * @param[in] dev          Device descriptor of the MFRC522
+ * @param[in] block_addr   MIFARE Classic: The block (0-0xff) number.
+ *                         MIFARE Ultralight: The page (2-15) to write to.
+ * @param[in] buffer       The 16 bytes to write to the PICC
+ * @param[in] buffer_size  Buffer size, must be at least 16 bytes. Exactly 16
+ *                         bytes are written.
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
+ */
+mfrc522_status_code_t mfrc522_mifare_write(mfrc522_t *dev,
+                                           uint8_t block_addr,
+                                           uint8_t *buffer,
+                                           uint8_t buffer_size);
+
+/**
+ * @brief Write a 4 byte page to the active MIFARE Ultralight PICC
+ *
+ * @param[in] dev          Device descriptor of the MFRC522
+ * @param[in] page         The page (2-15) to write to
+ * @param[in] buffer       The 4 bytes to write to the PICC
+ * @param[in] buffer_size  Buffer size, must be at least 4 bytes. Exactly 4
+ *                         bytes are written.
+ *
+ * @return MFRC522_STATUS_OK on success, MFRC522_STATUS_??? otherwise
+ */
+mfrc522_status_code_t mfrc522_mifare_ultralight_write(mfrc522_t *dev,
+                                                      uint8_t page,
+                                                      uint8_t *buffer,
+                                                      uint8_t buffer_size);
```

Comment:
I don't get why `buffer_size` is necessary, the user must provide always 4 bytes.

Reply:
I guess the upstream library wanted to extend on this some day, but as it is in complete feature freeze for years now I removed it.